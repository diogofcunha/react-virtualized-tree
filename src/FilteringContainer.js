import React from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';
import classNames from 'classnames';

import DefaultGroupRenderer from './filtering/DefaultGroupRenderer';
import {Node} from './shapes/nodeShapes';
import {filterNodes} from './selectors/filtering';

const indexByName = searchTerm => ({name}) => {
  const upperCaseName = name.toUpperCase();
  const upperCaseSearchTerm = searchTerm.toUpperCase();

  return upperCaseName.indexOf(upperCaseSearchTerm.trim()) > -1;
};

export default class FilteringContainer extends React.Component {
  state = {
    filterText: '',
    filterTerm: '',
  };

  getChildContext = () => {
    return {unfilteredNodes: this.props.nodes};
  };

  static childContextTypes = {
    unfilteredNodes: PropTypes.arrayOf(PropTypes.shape(Node)).isRequired,
  };

  static defaultProps = {
    debouncer: debounce,
    groupRenderer: DefaultGroupRenderer,
    indexSearch: indexByName,
  };

  constructor(props) {
    super(props);

    this.setFilterTerm = props.debouncer(this.setFilterTerm, 300);
  }

  setFilterTerm() {
    this.setState(ps => ({filterTerm: ps.filterText}));
  }

  handleFilterTextChange = e => {
    const filterText = e.target.value;

    this.setState({filterText});

    this.setFilterTerm();
  };

  render() {
    const {filterTerm, filterText} = this.state;
    const {
      nodes,
      children: treeRenderer,
      groups,
      selectedGroup,
      groupRenderer: GroupRenderer,
      onSelectedGroupChange,
      indexSearch,
    } = this.props;

    const relevantNodes =
      groups && selectedGroup && groups[selectedGroup]
        ? filterNodes(groups[selectedGroup].filter, nodes)
        : {nodes, nodeParentMappings: {}};

    const {nodes: filteredNodes, nodeParentMappings} = filterTerm
      ? filterNodes(indexSearch(filterTerm, relevantNodes.nodes), relevantNodes.nodes)
      : relevantNodes;

    return (
      <div className="tree-filter-container">
        <div className={classNames('tree-lookup-input', {group: !!groups})}>
          <input value={filterText} onChange={this.handleFilterTextChange} placeholder="Search..." />
          <i aria-hidden="true" className="mi mi-11 mi-search" />
          {groups && <GroupRenderer groups={groups} selectedGroup={selectedGroup} onChange={onSelectedGroupChange} />}
        </div>
        {treeRenderer({nodes: filteredNodes, nodeParentMappings})}
      </div>
    );
  }
}

FilteringContainer.propTypes = {
  children: PropTypes.func.isRequired,
  debouncer: PropTypes.func,
  groups: PropTypes.object,
  selectedGroup: PropTypes.string,
  groupRenderer: PropTypes.func,
  onSelectedGroupChange: PropTypes.func,
  indexSearch: PropTypes.func,
};
