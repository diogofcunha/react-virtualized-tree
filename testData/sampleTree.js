export const COLLAPSED_CHILDREN = {
  id: 3,
  name: 'Leaf 3',
  state: {
    expanded: false,
    favorite: true,
    deletable: true,
  },
  children: [
    {
      id: 'c-3',
      name: 'Leaf 3 Child',
      state: {},
    },
  ],
};

export const EXPANDED_CHILDREN = {
  id: 2,
  name: 'Leaf 2',
  state: {
    expanded: true,
    deletable: true,
  },
  children: [
    COLLAPSED_CHILDREN,
    {
      id: 4,
      name: 'Leaf 4',
    },
  ],
};

export const EXPANDED_NODE_IN_ROOT = {
  id: 0,
  name: 'Leaf 1',
  state: {
    expanded: true,
  },
  children: [
    EXPANDED_CHILDREN,
    {
      id: 5,
      name: 'Leaf 5',
    },
  ],
};

export const COLLAPSED_NODE_IN_ROOT = {
  id: 1,
  name: 'Leaf 6',
  state: {
    expanded: false,
    deletable: true,
  },
  children: [
    {
      id: 6,
      name: 'Leaf 7',
      state: {
        expanded: false,
      },
      children: [
        {
          id: 7,
          name: 'Leaf 8',
        },
        {
          id: 8,
          name: 'Leaf 9',
        },
      ],
    },
    {
      id: 9,
      name: 'Leaf 10',
    },
  ],
};

export const DELETABLE_IN_ROOT = {
  id: 'z',
  name: 'Leaf z',
  state: {
    deletable: true,
    favorite: true,
  },
};

export const DELETABLE_CHILDREN = EXPANDED_CHILDREN;

export const Nodes = [
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
];
