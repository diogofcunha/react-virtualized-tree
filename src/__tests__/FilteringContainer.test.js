import React from 'react';
import {render, cleanup} from 'react-testing-library';
import ReactTestUtils from 'react-dom/test-utils';

import FilteringContainer from '../FilteringContainer';
import {Nodes} from '../../testData/sampleTree';

describe('FilteringContainer', () => {
  afterEach(cleanup);

  const setup = (extraProps = {}) => {
    const child = jest.fn().mockImplementation(() => <div>Child</div>);

    const props = {nodes: Nodes, debouncer: v => v, ...extraProps};

    const wrapper = render(<FilteringContainer {...props}>{child}</FilteringContainer>);

    return {
      changeFilter: value => {
        const input = wrapper.getByPlaceholderText('Search...');

        ReactTestUtils.Simulate.change(input, {
          target: {
            value,
          },
        });
      },
      getInjectedNodes: () => {
        return child.mock.calls.slice(-1)[0][0].nodes;
      },
      wrapper,
      child,
      props,
    };
  };

  describe('filtering', () => {
    it('should not filter when searchTerm is empty', () => {
      const {changeFilter, props, getInjectedNodes} = setup();

      changeFilter('');

      expect(getInjectedNodes()).toEqual(props.nodes);
    });

    it('should filter for deepsearch', async () => {
      const {changeFilter, getInjectedNodes} = setup();

      changeFilter('2');

      expect(getInjectedNodes()).toMatchSnapshot();
    });

    it('should filter when results match more then 1 node', () => {
      const {changeFilter, getInjectedNodes} = setup();

      changeFilter('1');

      expect(getInjectedNodes()).toMatchSnapshot();
    });

    it('should filter when there are no results', () => {
      const {changeFilter, getInjectedNodes} = setup();

      changeFilter('Node');

      expect(getInjectedNodes()).toEqual([]);
    });

    it('should ignore boundarie spaces when filtering', () => {
      const {changeFilter, getInjectedNodes} = setup();

      changeFilter('1 ');

      expect(getInjectedNodes()).toMatchSnapshot();
    });

    it('should ignore casing when filtering', () => {
      const {changeFilter, getInjectedNodes} = setup();

      changeFilter('LEAf 3');

      expect(getInjectedNodes()).toMatchSnapshot();
    });
  });

  describe('groups', () => {
    it('when groups do not exist should not render groups related info', () => {
      const {wrapper} = setup();

      expect(wrapper.container.querySelector('.group')).toBeNull();
      expect(wrapper.container.querySelector('.tree-group')).toBeNull();
    });

    describe('when groups exist', () => {
      const EXPANDED = 'EXPANDED';

      const setupWithGroups = (extraProps = {}) =>
        setup({
          groups: {
            [EXPANDED]: {
              name: 'Expanded',
              filter: node => (node.state || {}).expanded,
            },
            FAVORITES: {
              name: 'Favorites',
              filter: node => (node.state || {}).favorite,
            },
          },
          selectedGroup: EXPANDED,
          onSelectedGroupChange: jest.fn(),
          ...extraProps,
        });

      it('should render the expected className', () => {
        const {wrapper} = setupWithGroups();

        expect(wrapper.container.querySelector('.group')).not.toBeNull();
      });

      it('should render the DefaultGroupRenderer when one is not injected as a prop', () => {
        const {wrapper} = setupWithGroups();

        expect(wrapper.container.querySelector('.tree-group')).toMatchSnapshot();
      });

      it('should render a injected groupRenderer', () => {
        const GroupRenderer = props => <div data-testid="group-renderer">Group renderer! {JSON.stringify(props)}</div>;

        const {wrapper} = setupWithGroups({
          groupRenderer: GroupRenderer,
        });

        expect(wrapper.getByTestId('group-renderer')).toMatchSnapshot();
      });

      it('should filter results based on the selected group', () => {
        const {getInjectedNodes} = setupWithGroups();

        expect(getInjectedNodes()).toMatchSnapshot();
      });

      it('should work when matched with filtering', () => {
        const {getInjectedNodes, changeFilter} = setupWithGroups();

        changeFilter('1');

        expect(getInjectedNodes()).toMatchSnapshot();
      });
    });
  });
});
