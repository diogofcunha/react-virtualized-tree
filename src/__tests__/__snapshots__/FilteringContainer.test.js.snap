// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilteringContainer filtering should filter for deepsearch 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
]
`;

exports[`FilteringContainer filtering should filter when results match more then 1 node 1`] = `
Array [
  Object {
    "children": Array [],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
]
`;

exports[`FilteringContainer filtering should ignore boundarie spaces when filtering 1`] = `
Array [
  Object {
    "children": Array [],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
]
`;

exports[`FilteringContainer filtering should ignore casing when filtering 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "id": "c-3",
                "name": "Leaf 3 Child",
                "state": Object {},
              },
            ],
            "id": 3,
            "name": "Leaf 3",
            "state": Object {
              "deletable": true,
              "expanded": false,
              "favorite": true,
            },
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
]
`;

exports[`FilteringContainer groups when groups exist should filter results based on the selected group 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
]
`;

exports[`FilteringContainer groups when groups exist should render a injected groupRenderer 1`] = `
<div
  data-testid="group-renderer"
>
  Group renderer! 
  {"groups":{"EXPANDED":{"name":"Expanded"},"FAVORITES":{"name":"Favorites"}},"selectedGroup":"EXPANDED"}
</div>
`;

exports[`FilteringContainer groups when groups exist should render the DefaultGroupRenderer when one is not injected as a prop 1`] = `
<select
  class="tree-group"
>
  <option
    value="EXPANDED"
  >
    Expanded
  </option>
  <option
    value="FAVORITES"
  >
    Favorites
  </option>
</select>
`;

exports[`FilteringContainer groups when groups exist should work when matched with filtering 1`] = `
Array [
  Object {
    "children": Array [],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
]
`;
